python/activator/middleware_interface.py

```diff
@@ -740,20 +741,42 @@ def run_pipeline(self, exposure_ids: set[int]) -> None:
             f" and exposure in ({','.join(str(x) for x in exposure_ids)})"
             " and visit_system = 0"
         )
-        pipeline = self._prep_pipeline()
-        executor = SeparablePipelineExecutor(self.butler, clobber_output=False, skip_existing_in=None)
-        qgraph = executor.make_quantum_graph(pipeline, where=where)
-        if len(qgraph) == 0:
+        # Try pipelines in order until one works.
+        for pipeline_file in self._get_pipeline_files():
+            try:
+                pipeline = self._prep_pipeline(pipeline_file)
+            except FileNotFoundError as e:
+                raise RuntimeError from e
+            init_output_run = self._get_init_output_run(pipeline_file, self._day_obs)
+            output_run = self._get_output_run(pipeline_file, self._day_obs)
+            executor = SeparablePipelineExecutor(
+                Butler(butler=self.butler,
+                       collections=[output_run, init_output_run] + list(self.butler.collections),
+                       run=output_run),
+                clobber_output=False,
+                skip_existing_in=None
+            )
+            qgraph = executor.make_quantum_graph(pipeline, where=where)
+            if len(qgraph) == 0:
+                # Diagnostic logs are the responsibility of GraphBuilder.
+                _log.error(f"Could not build quantum graph for {pipeline_file}; "
+                           "see previous logs for details.")
+                continue
+            # Past this point, partial execution creates datasets.
+            # Don't retry -- either fail (raise) or break.
```

Comment:
Do I understand correctly that for now, the only scenario to fallback is a failure in making the quantum graph?

Reply:
Yes, at least for now. Do you know of any other scenarios where this is a good strategy?

Reply:
Got another one that might be relevant here: quantum graph builds fine, but the needed datasets aren't really available so it fails. It can show as something like
`FileNotFoundError: Not enough datasets (0) found for non-optional connection`...

Reply:
I assume this happens only with `PrerequisiteInput`? I.e., it does not affect templates?

Reply:
Looks like only with `PrerequisiteInput`  
https://github.com/lsst/pipe_base/blob/main/python/lsst/pipe/base/connections.py#L852-L855

I haven't traced down the middleware enough to understand whether the qgraph would be built successfully in this case (I don't know when `adjustQuantum` happens).  When I encountered this error, I don't think the PP fallback pipeline ran, so I'd guess the qgraph was built.

Reply:
It looks like, despite the docstring, `adjustQuantum` is actually run [right before the quantum is executed](https://github.com/lsst/ctrl_mpexec/blob/8caecd393392dbc7d64e955c69b1c8bb9059ce55/python/lsst/ctrl/mpexec/singleQuantumExecutor.py#L431-L433).

Reply:
Ah, the qgraph building would fail. But it's not generating an empty qgraph which is the case the PP fallback pipeline is designed for.

Reply:
Sorry for the confusion on Slack! I've filed [DM-41596](https://jira.lsstcorp.org/browse/DM-41596) to address the bug.